Treat empty array as NULL; and fix value relation filter with NULL

elpaso opened the pull request:
Fix #46369: 46369 value relation null filtering
Fix #46366: not null constraint ignored when allowmulti

elpaso committed:
Fix drill-down filtering on new features.

Fix #46369

elpaso committed:
Fix constraint check when AllowMulti

Fix #46366

m-kuhn commented on src/core/vector/qgsvectorlayerutils.cpp:
```diff
@@ -431,9 +431,20 @@ bool QgsVectorLayerUtils::validateAttribute( const QgsVectorLayer *layer, const
-      valid = valid && !value.isNull();
 
-      if ( value.isNull() )
+      const bool allowMulti { field.editorWidgetSetup().config().value( QStringLiteral( "AllowMulti" ), false ).toBool() };
```
Can we move this into a `bool QgsFieldFormatter::validate()` (or similar) method?

elpaso replied:
maybe, with what kind of signature? The formatter has no knowledge of constraints.

m-kuhn replied:
`isNull()`?

gioman commented:
@elpaso thanks, thanks... thanks!

m-kuhn commented:
I'm half convinced by this approach.

1. Technically, an empty array is not null
This is what postgres tells me:

```sql
select '{}'::int[] is null;
 ?column? 
----------
 f
```

2. QGS Core should not be reading into widget configs. It should call methods on fieldformatter where each widget can deal with its config.
3. The approach taken in here relies on the array being stored in a text field with hstore formatting, it will not work if the underlying data is stored as a native array or json structure in the provider.

elpaso commented:
> I'm half convinced by this approach.
> 
>     1. Technically, an empty array is not null
>        This is what postgres tells me:
> 
> 
> ```sql
> select '{}'::int[] is null;
>  ?column? 
> ----------
>  f
> ```
> 
>     2. QGS Core should not be reading into widget configs. It should call methods on fieldformatter where each widget can deal with its config.
> 
>     3. The approach taken in here relies on the array being stored in a text field with hstore formatting, it will not work if the underlying data is stored as a native array or json structure in the provider.
> 
> 
> I could also see the following approaches:
> 
>     * Let the widget return NULL for empty arrays, that will work with the unmodified constraint validation logic plus it will match what a database validation does as well
> 
>     * Add a new "must not be empty" setting to the (multi enabled) widget, which will give hints on the gui
> 
>     * Hand this off to the user with a `array_length("field") > 0` function, but that won't work for text fields
> 
>     * Move this logic into a dedicated `bool QgsFieldFormatter::isNull(QVariant)` method, but as mentioned, that actually changes the meaning of null

I would not be much worried about what NULL exactly means for a DB in this context, I think that the user intention when she sets a NOT NULL constraint in these widgets doesn't leave much space to interpretations and can be read as "force the user to make a (multiple?) choice and do not allow a blank".

I will look into the widget returning invalid, it looks the most logical to me, thanks for the hints.

elpaso committed:
Revert "Fix constraint check when AllowMulti"

This reverts commit 355e3fb3fae492b406bb28688781ecaea0f05b9d.

elpaso committed:
Valu relation widget: fix allowMulti + not null

Fix #46366 by changing the returned value from the
widget when allowMulti is ON and allowNull is OFF,
in this case when the user makes no selection
an invalid QVariant is returned instead of a (valid)
empty list of QVariant(s).

elpaso commented:
@m-kuhn I think I've found the right approach with my last commit.

Of course from a UX the situation is less than ideal, see the various mentions to `NULL` an `null` in the widget config plus the conflicting options to allow `NULL` in the widget and set not `null` (notice the case) in the constraints section.

Anyway, this is a small fix for the value-relation only, I guess we should check the other widgets as well.

m-kuhn reviewed (approved):
Looks good now, thanks.
Agreed, the null handling (and consistent spelling) could be improved.
Also it would be nice to change the allow null text to allow empty if allow multi is checked.
 Cc @DelazJ

DelazJ commented:
@m-kuhn sorry for the delay. What was the question, please?

m-kuhn commented:
I thought the topic of wording on the config UI discussed in https://github.com/qgis/QGIS/pull/46439#issuecomment-991581012 could be interesting for you
